### PR TITLE
Draft: add typed mmap to improve code safety

### PR DESCRIPTION
**Pull request by timvisee**
Builds upon <https://github.com/qdrant/qdrant/pull/1838> to add safer mmap usage.

This PR adds a typed mmap struct to directly map a type, slice or bitslice on a memory mapped file. The memory mapped type `MmapType<T>` can be directly used as if it were type `T`.

This is a simplified and generic interface for using types on memory mapped files. It moves all the potentially dangerous logic into a single module (`mmap_type.rs`). This makes it easier to review, check for soundness and to reason about.

All existing usages of `MmapMut` (and `unsafe`), that were spread over the codebase, have now been replaced with this unified interface.

For example:

```rust
// Open mmap and map to Status type
let mmap = open_write_mmap(status_file)?;
let status: MmapType<Status> = MmapType::from(mmap);

// We can now directly use Status
println!("{}", status.len);
println!("{}", status.current_file_id);
```

### Open tasks
- [x] Panic-free interface?
- [ ] Check for soundness
  - [x] Check Rust's aliasing rules
- [x] Memory map alignment on Windows

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by generall — APPROVED** [review]

File: lib/segment/src/vector_storage/mmap_type.rs [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+        "header not multiple of size T",
+    );
+    assert_eq!(
+        bytes.len() % mem::size_of::<T>(),
+        0,
+        "bytes not multiple of size T",
+    );
+
+    // Transmute slice types
+    slice::from_raw_parts_mut(bytes.as_ptr() as *mut T, bytes.len() / mem::size_of::<T>())
```
Comment (Jesse-Bakker) [inline lib/segment/src/vector_storage/mmap_type.rs:359]: This is UB under stacked borrows: &[u8] -> *const u8 -> *mut T -> &mut[T].
Using `bytes.as_mut_ptr()` instead of `bytes.as_ptr()` should be sound though.

Equivalent UB in `mmap_to_type_unbounded`, also fixed by using `as_mut_ptr` instead of `as_ptr`

File: lib/segment/src/vector_storage/mmap_type.rs [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+/// Assert slice `&[S]` is correctly aligned for type `T`.
+///
+/// # Panics
+///
+/// Panics when alignment is wrong.
+fn assert_alignment<S, T>(bytes: &[S]) {
+    // We do not enforce alignment on Windows at this time, because the Windows API doesn't
+    // guarantee memory mapped segments to be aligned.
+    //
+    // We should look into this again to confirm this is sound.
```
Comment (Jesse-Bakker) [inline lib/segment/src/vector_storage/mmap_type.rs:371]: Producing a reference to unaligned data, even if not reading or writing through that reference, is UB: https://doc.rust-lang.org/reference/behavior-considered-undefined.html

File: lib/segment/src/vector_storage/mmap_type.rs (hunk 1 above)
Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: Good catch! Thanks a lot.

File: lib/segment/src/vector_storage/mmap_type.rs (hunk 2 above)
Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: This is a hard problem. We discovered that mmap'ed files are not aligned on Windows.

We discussed this, and didn't find an easy solution to mitigate this. The very same is already in other places. We decided to match that for now, which means, disabling this check on Windows, until we find a better solution.

Do you have a good suggestion for this? We might even end up disabling mmap support on Windows+ARM builds.

Comment (Jesse-Bakker) [reply lib/segment/src/vector_storage/mmap_type.rs]: Windows (at least newer versions) has API's to create aligned mmaps and in the cases I can see, it by default aligns to 64kB boundaries on my system. Which cases did you find where the map is not aligned? Is it only on ARM?

Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: Thanks for your reply.

On Windows it fails in unit tests with a very basic alignment check.

### What happens

This is already merged in the `dev` branch, so we can use that as base.

If we enable the alignment check on Windows here:

https://github.com/qdrant/qdrant/blob/83f8b50acadd25d47a9bdced1e762b924a7b272d/lib/segment/src/common/mmap_type.rs#L463-L472

This simple check fails (amongst others):

https://github.com/qdrant/qdrant/blob/83f8b50acadd25d47a9bdced1e762b924a7b272d/lib/segment/src/common/mmap_type.rs#L498

Here's the error on CI when it was enabled: https://github.com/qdrant/qdrant/actions/runs/4923938726/jobs/8796365496#step:6:1213

If you enable the assertion on Windows as noted above, you can probably easily run the test with: `cargo test --profile dev --package segment test_open-zero`

All it does is:
- Create file on disk (with various sizes)
- Open the file (read/write)
- Open mmap with [`MmapMut::map_mut`](https://docs.rs/memmap2/latest/memmap2/struct.MmapMut.html#method.map_mut)
- Then our alignment assertion complains the bytes in memory are off-by-one (`bytes.as_ptr().align_offset(mem::align_of::<T>()) == 7`)

Either the file isn't mapped in a 64k grid, or the check is somehow faulty. I don't have good knowledge of Windows APIs.

---

> Is it only on ARM?

This is on CI. Which is x86_64.

Comment (Jesse-Bakker) [reply lib/segment/src/vector_storage/mmap_type.rs]: Thanks for the detailed reply. I'm intrigued, so I'll look into it a bit and let you know what I find. I have a feeling the actual windows syscall returns an aligned pointer and some pointer arithmetic in memmap2 makes it unaligned afterwards, but for the moment that feeling is completely unsubstantiated.

Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: That would be fantastic! Let me know if I can be of any help. I'd be happy to prepare a more concise example in code for you.

Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: That makes a lot of sense. Thanks a lot for taking the time to investigate this and for figuring this out. That's super useful! A bit of a bummer that the crate doesn't return this edge case as an error, that would have made debugging a lot easier.

I'll fix this tomorrow, adding a check to mitigate this very situation.

It makes sense it fails, because in the test we actually create a file with `.set_len(0)`.

Thanks once again!

File: lib/segment/src/vector_storage/mmap_type.rs (hunk 1 above)
Comment (Jesse-Bakker) [reply lib/segment/src/vector_storage/mmap_type.rs]: Just wanted to mention I found this by changing MmapMut to Vec and running the tests with miri. 
Just a cool trick that may be useful in the future. Also, the rest of the logic was sound according to miri, so that might be a nice assurance.

Comment (timvisee) [reply lib/segment/src/vector_storage/mmap_type.rs]: :pray:
